libraries/AP_Telemetry/AP_Telemetry.h

```diff
@@ -0,0 +1,85 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Notify/AP_Notify.h>
+#include <AP_HAL/utility/RingBuffer.h>
+
+#define TELEM_PAYLOAD_STATUS_CAPACITY          5 // size of the message buffer queue (max number of messages waiting to be sent)
+
+// for fair scheduler
+#define TELEM_TIME_SLOT_MAX               15
+
+class AP_Telemetry {
+public:
+    AP_Telemetry(uint8_t time_slots) : _time_slots(time_slots) {}
+    virtual ~AP_Telemetry() {};
+
+    /* Do not allow copies */
+    AP_Telemetry(const AP_Telemetry &other) = delete;
+    AP_Telemetry &operator=(const AP_Telemetry&) = delete;
+
+    // add statustext message to message queue
+    void queue_message(MAV_SEVERITY severity, const char *text);
+
+    // update error mask of sensors and subsystems. The mask uses the
+    // MAV_SYS_STATUS_* values from mavlink. If a bit is set then it
+    // indicates that the sensor or subsystem is present but not
+    // functioning correctly
+    uint32_t sensor_status_flags() const;
+
+protected:
+    void passthrough_wfq_adaptive_scheduler();
```

Comment:
Andy since you're refactoring I guess we can use a more generic name like wfq_scheduler() and get rid of the passtrough legacy, you just proved it's a generic tool :-)
Same for setup_passthrough_wfq() to setup_wfq_scheduler(), and lastly setup_passthrought() to a simple setup()

Reply:
Perfect - will do!

Reply:
Ok I have slightly modifed this - let me know if you agree with what I have done